text/3537-msrv-resolver.md

```diff
@@ -0,0 +1,565 @@
+- A `Cargo.lock` should not resolve differently when upgrading Rust without any other action.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## The `rust-version` field
+
+*(update to [manifest documentation](https://doc.rust-lang.org/cargo/reference/manifest.html#the-rust-version-field))*
+
+The `rust-version` field is an optional key that tells cargo what version of the
+Rust language and compiler your package can be compiled with. If the currently
+selected version of the Rust compiler is older than the stated version, cargo
+will exit with an error, telling the user what version is required.
+To support this, Cargo will prefer dependencies that are compatible with your `rust-version`.
+
+The first version of Cargo that supports this field was released with Rust 1.56.0.
+In older releases, the field will be ignored, and Cargo will display a warning.
+
+```toml
+[package]
+# ...
+rust-version = "1.56"
+```
+
+The Rust version must be a bare version number with two or three components; it
+cannot include semver operators or pre-release identifiers. Compiler pre-release
+identifiers such as -nightly will be ignored while checking the Rust version.
+The `rust-version` must be equal to or newer than the version that first
+introduced the configured `edition`.
+
+The `rust-version` may be ignored using the `--ignore-rust-version` option.
+
+Setting the `rust-version` key in `[package]` will affect all targets/crates in
+the package, including test suites, benchmarks, binaries, examples, etc.
+
+## Rust Version
+
+*(update to [Dependency Resolution's Other Constraints documentation](https://doc.rust-lang.org/cargo/reference/resolver.html))*
+
+When multiple versions of a dependency satisfy all version requirements,
+cargo will prefer those with a compatible `package.rust-version` over those that
+aren't compatible.
+Some details may change over time though `cargo check && rustup update && cargo check` should not cause `Cargo.lock` to change.
+
+#### `build.resolver.precedence`
+
+*(update to [Configuration](https://doc.rust-lang.org/cargo/reference/config.html))*
+
+* Type: string
+* Default: "rust-version=package"
+* Environment: `CARGO_BUILD_RESOLVER_PRECEDENCE`
+
+Controls how `Cargo.lock` gets updated on changes to `Cargo.toml` and with `cargo update`.  This does not affect `cargo install`.
+
+* `maximum`: Prefer the highest compatible versions of dependencies
+* `minimum`: Prefer the lowest versions of dependencies
+* `rust-version=package`: Prefer dependencies where their `rust-version` is compatible with `package.rust-version`
+* `rust-version=rustc`: Prefer dependencies where their `rust-version` is compatible with `rustc --version`
+* `rust-version=<X>[.<Y>[.<Z>]]`: Prefer dependencies where their `rust-version` is compatible with the specified version
+
+`rust-version=` values can be overridden with `--ignore-rust-version` which will fallback to `maximum`.
```

Comment:
@joshtriplett 
> 👍 for introducing the MSRV-aware resolver.
> 
> I'm hesitant to introduce all the variations on the configuration option. In particular, I'm not sure we should ever stabilize `minimum`, or for that matter `rust-version=1.X.Y`. I'm also not sure we ever want `rust-version=rustc`; that seems like it'd encourage people to try a combination of dependencies that the package maintainer hasn't tested, that _also_ aren't the latest versions so nobody else is likely to have tested them either.
> 
> I'd be inclined to _just_ have `maximum` and `rust-version=package` (with the latter perhaps having a shorter name if we don't need as many variations).
> 
> Could I request moving the remaining variations into future work, such that the RFC is only proposing the `maximum` and `rust-version=package` options? And, potentially, consider proposing a shorter name for the latter, such as `msrv`?

Reply:
I hope its ok that I moved this as I felt this might be enough discussion to warrant a thread to make it easier to follow  with other discussions.

`minimal` was listed as "unstable" and `rust-version=X.Y.Z` as "future".  I did those inline to make potential use cases for this field clearer to help in the discussion of naming.

I can see deferring `rust-version=rustc`.  I know some people are advocating for it as the default but if we're prioritizing the single-MSRV workflow, then the value add of it is small compared to `maximum`.  It is the behavior that would be useful for `cargo install` (even if we want to expose that as a "preview" by not changing the default behavior but allowing it to be overridden).

Reply:
c668698 simplified things and cleared things up.

Reply:
@epage The qualifiers you added, together with removing the variations from the earlier mention, address my concern.

I'm not *yet* sure if `rust-version=rustc` is going to be the right answer for `cargo install`. I'd be more inclined to make `cargo install --locked` the default, which *should* solve the problem.

Reply:
The big problem for `cargo install` is selecting the right version upfront.  Maybe that'd be a different config field.

Reply:
I conflated top-level package selection with dependency resolution.  That was fixed in 5c55bd1.